### PR TITLE
NRF: add UART implementation

### PR DESCRIPTION
**Pull request by deadprogram**
This PR adds the UART interface's implementation based on code as used in the AVR machine type.

### CODE REVIEW

File: src/machine/machine_nrf.go [hunk 1]
```diff
@@ -35,3 +37,168 @@ func (p GPIO) Set(high bool) {
...
+// WriteByte writes a byte of data to the UART.
+func (uart UART) WriteByte(c byte) error {
+	nrf.UART0.EVENTS_TXDRDY = 0
+	nrf.UART0.TXD = nrf.RegValue(c)
+	for nrf.UART0.EVENTS_TXDRDY == 0 {
+	}
+	return nil
+}
+
+// Buffered returns the number of bytes current stored in the RX buffer.
```
Comment (aykevl) [inline src/machine/machine_nrf.go]: Nit: spelling (current vs currently).

Comment (deadprogram) [reply src/machine/machine_nrf.go]: Fixed.

File: src/machine/machine_nrf.go [hunk 2]
```diff
@@ -35,3 +37,168 @@ func (p GPIO) Set(high bool) {
 func (p GPIO) Get() bool {
 	return (nrf.P0.IN>>p.Pin)&1 != 0
 }
+
+// UART
+type UARTConfig struct {
+	Baudrate uint32
```
[the remark this answers is not in the document]
Comment (deadprogram) [reply src/machine/machine_nrf.go]: OK will leave to future iteration.

File: src/machine/machine_nrf.go [hunk 3]
```diff
@@ -35,3 +37,168 @@ func (p GPIO) Set(high bool) {
...
+	if bufferUsed() != 0 {
+		tail++
+		return byte(rxbuffer[tail%bufferSize])
+	}
+	return 0
+}
+
+//go:export UARTE0_UART0_IRQHandler
+func handleUART0() {
+	if nrf.UART0.EVENTS_RXDRDY == 1 {
```
[the remark this answers is not in the document]
Comment (deadprogram) [reply src/machine/machine_nrf.go]: Every instruction helps. Updated.

File: src/machine/machine_nrf.go [hunk 4]
```diff
@@ -35,3 +37,168 @@ func (p GPIO) Set(high bool) {
...
+	nrf.UART0.TASKS_STARTTX = 1
+	nrf.UART0.TASKS_STARTRX = 1
+	nrf.UART0.INTENSET = nrf.UART_INTENSET_RXDRDY_Msk
+
+	// Enable RX IRQ.
+	arm.EnableIRQ(nrf.IRQ_UARTE0_UART0)
+}
+
+// SetBaudRate sets the communication speed for the UART.
+func (uart UART) SetBaudRate(br uint32) {
```
[the remark this answers is not in the document]
Comment (deadprogram) [reply src/machine/machine_nrf.go]: I went with `BaudRate` then. Updated.

**Comment by deadprogram** [comment]
Created new file `uart.go` under `/machine` directory, and modified both NRF and AVR implementations to share the common code.

**Comment by deadprogram** [comment]
I added the `BaudRate` calculation formula as you suggested @aykevl instead of the previous switch statement.

File: src/machine/board_pca10040.go [hunk 5]
```diff
@@ -19,3 +19,9 @@ const (
 	BUTTON3 = 15
 	BUTTON4 = 16
 )
+
+// UART pins for NRF52840-DK
+const (
+	UART_TX_PIN = 6
+	UART_RX_PIN = 8
```
[the remark this answers is not in the document]
Comment (aykevl) [reply src/machine/board_pca10040.go]: You are right, the naming convention is different from what I thought. Not what I would have chosen but if that's the convention...
https://stackoverflow.com/questions/22688906/go-naming-conventions-for-const

Maybe it is best to fix all pin names in a single commit, after this commit is merged.

[the remark this answers is not in the document]
Comment (deadprogram) [reply src/machine/board_pca10040.go]: That seems like a good plan.

**Comment by deadprogram** [comment]
Any action on this PR? I have another PR to add ADC that should have this one land first, probably...

**Comment by aykevl** [comment]
With this update, the PR breaks the bluepill target and `make run-blinky2`:

```
~/src/github.com/aykevl/tinygo$ make flash-blinky1 TARGET=bluepill
go build -o build/tgo -i .
./build/tgo build -target bluepill -size=short -o build/blinky1.elf examples/blinky1
/home/ayke/src/github.com/aykevl/tinygo/src/machine/uart.go:37:8: invalid operation: uart (variable of type UART) has no field or method WriteByte
error: couldn't load packages due to errors: machine
Makefile:89: recipe for target 'build/blinky1.elf' failed
make: *** [build/blinky1.elf] Error 1
```

I have added a check to Travis-CI to automatically detect these kinds of bugs, so when you push a new commit it should be tested automatically.

I would suggest adding `// +build avr,nrf` build tags to uart.go.

**Comment by deadprogram** [comment]
The syntax is actually `// +build avr nrf` here, meaning "avr OR nrf". Change made and pushed.

**Comment by aykevl** [comment]
You're right, the comma says 'and' and it must be 'or'.
Merged in ef2ac095611dd857327074e1e83eaf9d9d3bcc22.

**Pull request closed without merging**